src/docs/rust.mdx

```diff
@@ -0,0 +1,323 @@
+```rust
+format_code_in_doc_comments = true
+```
+
+### Writing Tests
+
+All test functions should start with `test_` and contain the function name + a simple condition in their name. The test name should be as concise as possible and avoid redundant words (such as "should", "that"). Examples:
+
+- `test_parse_empty`
+- `test_parse_null`
+
+Write unit tests in a `tests` submodule in code.
+
+```rust
+fn foo() {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_foo() { .. }
+}
+```
+
+Integration tests should go into the `tests/` folder, ideally into a file based on the functionality they test. Such tests can only use the public interface.
+
+For libraries, consider providing examples in `examples/`.
+
+### Iterators
+
+Prefer writing explicit iterator types over `impl Iterator`. This allows to name the type in places like associated types or globals. The type name should end with `Iter` per naming convention.
+
+In addition to the standard `Iterator` trait, always consider to implement additional traits from `std::iter`:
+
+- `FusedIterator` in all cases unless there is a strong reason not to
+- `DoubleEndedIterator` if reverse iteration is possible
+- `ExactSizeIterator` if the size is known beforehand
+
+If it is exceptionally hard to write a custom iterator, then it can also be a private [newtype](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html) around a boxed iterator:
+
+```rust
+pub struct FooIter(Box<dyn Iterator<Item = Foo>>);
+
+impl Iterator for FooIter {
+    type Item = Foo;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+```
+
+### Async-await
+
+In general we like to migrate to async-await using code, though much code exists which does not yet do this.  During migration you may need normal functions which return futures, for their signature you should generally prefer `-> impl Future<Output = ...>` over other return types.  This allows the compiler to figure out the details for callsites and avoids having to force things to be pinned or not.  Pinning should only be done explicitly when part of fields in structs or similar.
+
+In **traits** you can not yet use `async fn` ([see this blog post](https://smallcultfollowing.com/babysteps/blog/2019/10/26/async-fn-in-traits-are-hard/)), in this case we have an exception and do the same as the [async-trait crate](https://crates.io/crates/async-trait) does and functions return `-> Pin<Box<dyn Future<Output = ...>>>`.  This allows seamless use in async-await code or in code dealing directly with futures.  Explicitly writing out the signatures is preferable over pulling in the macro syntactic sugar.
+
+### Discourage unwrap()
+
+This may seem obvious, but *almost always* discourage the use of `.unwrap()`. Even if a piece of code is guaranteed not to panic in its current form (because some precondition is met), it might be reused or refactored in a future situation where the precondition does not hold anymore.
+
+
```

Comment:
```suggestion
### Use `.get()` instead of slice syntax

Slice syntax (`&foo[a..b]`) will panic if the indices are out of bounds or out of order. Especially when dealing with untrusted input, it is better to use `.get(a..b)` and an `if let Some(...) =` expression.

### Use checked math

Arithmetic under/overflows will panic in debug builds, but will lead to wrong results or panics in other parts of the code (like slice syntax mentioned above) in release builds.
It is a good idea to always use checked math, such as `checked_sub` or `saturating_sub`. The `saturating` variants will be capped at the appropriate `MIN/MAX` of the underlying data type.
```